### PR TITLE
Add getIfAvailable API (#4956)

### PR DESCRIPTION
**Pull request by Kyoungwoong**
Motivation:

The current implementation of RequestLog access in AbstractHttpResponseHandler is verbose and potentially inefficient. Each operation checks the availability of RequestLog properties separately, accessing a volatile variable and computing the available flags multiple times. Additionally, the code style does not align with efficient and concise Kotlin programming practices.

Modifications:

Introduce new API for accessing RequestLog only when specific flags are available
Provide Kotlin users with the ability to utilize null-safe calls

Result:

Improved efficiency of RequestLog access by reducing redundant checks and computations
Enhanced compatibility for Kotlin developers by enabling null-safe calls when accessing RequestLog properties

### CODE REVIEW

**Comment by jrhee17** [comment]
> The current implementation of RequestLog access in AbstractHttpResponseHandler is verbose and potentially inefficient. Each operation checks the availability of RequestLog properties separately, accessing a volatile variable and computing the available flags multiple times

I think I can agree with the motivation of this PR 👍   Can you take a look at the lint failures?

File: core/src/main/java/com/linecorp/armeria/common/logging/DefaultRequestLog.java [hunk 1]
```diff
@@ -1521,6 +1534,19 @@ public boolean isAvailable(Iterable<RequestLogProperty> properties) {
             return true;
         }
 
+        @Nullable
+        @Override
+        public RequestLog getIfAvailable(RequestLogProperty... properties){
+            return isAvailable(properties) ? this : null;
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/common/logging/DefaultRequestLog.java:1540]: `isAvailable` is always true.

File: core/src/main/java/com/linecorp/armeria/common/logging/RequestLogAccess.java [hunk 2]
```diff
@@ -76,6 +76,18 @@ public interface RequestLogAccess {
      */
     boolean isAvailable(Iterable<RequestLogProperty> properties);
 
+    /**
+     * Returns {@link RequestLog} if all of the specified {@link RequestLogProperty}s are available.
+     */
+    @Nullable
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/common/logging/RequestLogAccess.java:82]:
```suggestion
    @UnstableApi
    @Nullable
```

File: core/src/main/java/com/linecorp/armeria/common/logging/RequestLogAccess.java [hunk 3]
```diff
@@ -76,6 +76,18 @@ public interface RequestLogAccess {
...
+    RequestLog getIfAvailable(RequestLogProperty... properties);
+
+    /**
+     * Returns {@link RequestLog} if all of the specified {@link RequestLogProperty}s are available.
+     */
+    @Nullable
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/common/logging/RequestLogAccess.java:88]:
```suggestion
    @UnstableApi
    @Nullable
```

File: core/src/main/java/com/linecorp/armeria/common/logging/DefaultRequestLog.java (hunk 1 above)
Comment (Kyoungwoong) [reply core/src/main/java/com/linecorp/armeria/common/logging/DefaultRequestLog.java:1540]: Should I return the RequestLog class without using the isAvailable function?

**Comment by Kyoungwoong** [comment]
I will resolve all comments. PTAL

File: core/src/test/java/com/linecorp/armeria/common/logging/DefaultRequestLogTest.java [hunk 4]
```diff
@@ -493,4 +493,19 @@ void toStringWithChildren() {
         assertThat(lines[2]).matches("^\\t\\{Request: \\{.*}, Response: \\{.*}}$");
         assertThat(lines[3]).matches("^\\t\\{Request: \\{.*}, Response: \\{.*}}$");
     }
+
+    @Test
+    void testGetIfAvailable(){
+        // Given
+        final ServiceRequestContext ctx = ServiceRequestContext.of(HttpRequest.of(HttpMethod.GET, "/"));
+        final DefaultRequestLog log = (DefaultRequestLog) ctx.log();
```
Comment (jrhee17) [inline core/src/test/java/com/linecorp/armeria/common/logging/DefaultRequestLogTest.java:501]: Question) Is there a need to cast this?

```suggestion
        final RequestLogAccess log = ctx.log();
```

**Review by jrhee17 — APPROVED** [review]
Left a minor comment, but looks good to me 👍    Please also resolve the lint failures

File: core/src/main/java/com/linecorp/armeria/common/logging/DefaultRequestLog.java [hunk 5]
```diff
@@ -535,33 +547,33 @@ private void propagateRequestSideLog(RequestLogAccess child) {
         // Update the available properties always by adding a callback,
         // because the child's properties will never be available immediately.
         child.whenAvailable(RequestLogProperty.REQUEST_START_TIME)
-             .thenAccept(log -> startRequest(log.requestStartTimeNanos(), log.requestStartTimeMicros()));
+                .thenAccept(log -> startRequest(log.requestStartTimeNanos(), log.requestStartTimeMicros()));
```
Comment (jrhee17) [inline core/src/main/java/com/linecorp/armeria/common/logging/DefaultRequestLog.java:550]: Did you set up your IDE as recommended? https://armeria.dev/community/developer-guide#setting-up-your-ide

Can you revert the indentation changes made in this file?

Comment (Kyoungwoong) [reply core/src/main/java/com/linecorp/armeria/common/logging/DefaultRequestLog.java:550]: I'm sorry. I'll fix it quickly.

**Comment by Kyoungwoong** [comment]
I got some problem. Please give some time.

**Review by trustin — APPROVED** [review]
Thank you, @Kyoungwoong! :bow::+1:

**Comment by minwoox** [comment]
Could you search and use this API where applicable, please?

**Comment by Kyoungwoong** [comment]
Yes, I'll do it

**Review by ikhoon — APPROVED** [review]
Looks nice! Thanks, @Kyoungwoong! 🚀❤️

**Comment by Kyoungwoong** [comment]
I'm sorry for my inexperience in this work. It was a good experience. Thank you.

**Comment by minwoox** [comment]
@Kyoungwoong Could you check the lint failure? 😉 
https://github.com/line/armeria/actions/runs/5526782829/jobs/10187668968?pr=4966

**Pull request merged**